### PR TITLE
feat(edgeless): support copy as PNG

### LINKED ISSUE #2442: Support `Copy as PNG` in the component toolbar

### PR DESCRIPTION
**Pull request by fundon**
Closes #2442

https://github.com/toeverything/blocksuite/assets/27926/42060d7d-5b07-42a2-8015-458e9fdea804

### CODE REVIEW

File: packages/blocks/src/__internal__/clipboard/edgeless-clipboard.ts [hunk 1]
```diff
@@ -321,4 +329,88 @@ export class EdgelessClipboard implements Clipboard {
...
+
+    try {
+      // waiting for canvas to render
+      await new Promise(requestAnimationFrame);
+
+      const blob = await html2Image.toBlob(container, {
+        cacheBust: true,
+      });
+      assertExists(blob);
+      await navigator.clipboard.write([
```
[the remark this answers is not in the document]
Comment (fundon) [reply packages/blocks/src/__internal__/clipboard/edgeless-clipboard.ts:416]: This is acceptable and allows users to make their own choices. `performNativeCopy` just copies the text content, and unable to paste to tldraw, excalidraw.

File: packages/blocks/src/__internal__/clipboard/edgeless-clipboard.ts [hunk 2]
```diff
@@ -321,4 +329,88 @@ export class EdgelessClipboard implements Clipboard {
...
+    const vx = cx - width / 2 / zoom;
+    const vy = cy - height / 2 / zoom;
+
+    const container = document.createElement('div');
+    container.style.position = 'relative';
+    container.style.width = `${width}px`;
+    container.style.height = `${height}px`;
+    _edgeless.appendChild(container);
+
+    if (framesLen) {
```
[the remark this answers is not in the document]
Comment (fundon) [reply packages/blocks/src/__internal__/clipboard/edgeless-clipboard.ts:367]: `transPageToCanvas` method will copy all page, but in this scene is a copy the selected shapes and frames to the clipboard.

File: packages/blocks/src/__internal__/clipboard/edgeless-clipboard.ts [hunk 3]
```diff
@@ -321,4 +329,88 @@ export class EdgelessClipboard implements Clipboard {
...
+      const renderer = new Renderer();
+      renderer.load(shapes);
+      renderer.setCenter(cx, cy);
+      renderer.setZoom(zoom);
+      renderer.attach(container);
+    }
+
+    try {
+      // waiting for canvas to render
+      await new Promise(requestAnimationFrame);
```
[the remark this answers is not in the document]
Comment (fundon) [reply packages/blocks/src/__internal__/clipboard/edgeless-clipboard.ts:402]: https://github.com/toeverything/blocksuite/blob/ade674af2b68a406f37d3dca2b2c423b01e41d3e/packages/phasor/src/renderer.ts#L253

Maybe `_shouldUpdate` should be public.

**Comment by fundon** [comment]
Waiting upstream #3216

**Comment by fundon** [comment]
@zuoxiaodong0815 This question was moved to the AFFiNE side?

**Comment by zuoxiaodong0815** [comment]
Sorry,  I have a misunderstanding of our communication. Please feel free to deal this issue. @fundon

**Comment by fundon** [comment]
Related to: https://github.com/toeverything/AFFiNE/pull/2932

**Review by zuoxiaodong0815 — APPROVED** [review]

**Pull request merged**